hoomd/md/EvaluatorBondTether.h

```diff
@@ -0,0 +1,181 @@
+/*! The parameters are:
+    - \a k_b (param.x) Bond stiffness
+    - \a l_min (param.y) minimum bond length
+    - \a l_c1 (param.w) cutoff length of repulsive part
+    - \a l_c0 (param.z) cutoff length of attractive part
+    - \a l_max (param.a) maximum bond length
+*/
+class EvaluatorBondTether
+    {
+    public:
+    //! Define the parameter type used by this bond potential evaluator
+    typedef tether_params param_type;
+
+    //! Constructs the pair potential evaluator
+    /*! \param _rsq Squared distance between the particles
+        \param _params Per type pair parameters of this potential
+    */
+    DEVICE EvaluatorBondTether(Scalar _rsq, const param_type& _params)
+        : rsq(_rsq), k_b(_params.k_b), l_min(_params.l_min), l_c1(_params.l_c1), l_c0(_params.l_c0),
+          l_max(_params.l_max)
+        {
+        }
+
+    //! This evaluator doesn't use diameter
+    DEVICE static bool needsDiameter()
+        {
+        return false;
+        }
+
+    //! Accept the optional diameter values
+    /*! \param da Diameter of particle a
+        \param db Diameter of particle
+    */
+    DEVICE void setDiameter(Scalar da, Scalar db) { }
+
+    //! Tether doesn't use charge
+    DEVICE static bool needsCharge()
+        {
+        return false;
+        }
+
+    //! Accept the optional charge values
+    /*! \param qa Charge of particle a
+        \param qb Charge of particle b
+    */
+    DEVICE void setCharge(Scalar qa, Scalar qb) { }
+
+    //! Evaluate the force and energy
+    /*! \param force_divr Output parameter to write the computed force divided by r.
+        \param bond_eng Output parameter to write the computed bond energy
+
+        \return True if they are evaluated or false if the bond
+                energy is not defined
+    */
+    DEVICE bool evalForceAndEnergy(Scalar& force_divr, Scalar& bond_eng)
+        {
+        Scalar r = sqrt(rsq);
+        Scalar U_att = Scalar(0.0);
+        Scalar F_att = Scalar(0.0);
+        Scalar U_rep = Scalar(0.0);
+        Scalar F_rep = Scalar(0.0);
+
+        if (r > l_c0)
+            {
+            U_att = k_b * (exp(Scalar(1.0) / (l_c0 - r)) / (l_max - r));
+            F_att = k_b
+                    * (((r - l_max) * exp(Scalar(1.0) / (l_c0 - r)) / (l_c0 - r) / (l_c0 - r)
+                        - exp(Scalar(1.0) / (l_c0 - r)))
+                       / (l_max - r) / (l_max - r));
+            }
+
+        if (r < l_c1)
+            {
+            U_rep = k_b * (exp(Scalar(1.0) / (r - l_c1)) / (r - l_min));
+            F_rep = k_b
+                    * (((r - l_min) * exp(Scalar(1.0) / (r - l_c1)) / (r - l_c1) / (r - l_c1)
+                        + exp(Scalar(1.0) / (r - l_c1)))
+                       / (r - l_min) / (r - l_min));
+            }
+        if (k_b != Scalar(0.0))
+            {
+            force_divr = (F_att + F_rep) / r;
+
+            bond_eng = U_att + U_rep;
+            }
+
+        return true;
```

Comment:
You can return `false` when the particle goes outside the valid range for the bond to get an error message.